File: EnergyAPI/src/main/java/fr/traqueur/energylib/api/types/EnergyType.java
package fr.traqueur.energylib.api.types;

import java.util.ArrayList;
import java.util.List;

/**
 * Represents an energy type.
 */
public interface EnergyType {

    /**
     * The default energy type.
     */
    List<EnergyType> TYPES = new ArrayList<>(List.of(EnergyTypes.values()));

    /**
     * Registers a new energy type.
     * @param type the type to register
     */
    static void registerType(EnergyType type) {
        if(TYPES.stream().anyMatch(t -> t.getName().equalsIgnoreCase(type.getName()))) {
            throw new IllegalArgumentException("EnergyType with name " + type.getName() + " already exists!");
        }
        TYPES.add(type);
    }

    /**
     * Gets the name of the energy type.
     * @return the name of the energy type
     */
    String getName();

}

File: FluxAPI/src/main/kotlin/io/github/Jadiefication/redstoneflux/api/types/EnergyType.kt
package io.github.Jadiefication.redstoneflux.api.types

/**
 * Represents an energy type.
 */
interface EnergyType {
    /**
     * Gets the name of the energy type.
     * @return the name of the energy type
     */
    val name: String?

    companion object {
        /**
         * Registers a new energy type.
         * @param type the type to register
         */
        fun registerType(type: EnergyType) {
            require(
                !TYPES.stream().anyMatch { t: EnergyType? ->
                    t!!.name.equals(
                        type.name,
                        ignoreCase = true,
                    )
                },
            ) { "EnergyType with name " + type.name + " already exists!" }
            TYPES.add(type)
        }

        /**
         * The default energy type.
         */
        val TYPES: MutableList<EnergyType?> =
            ArrayList<EnergyType?>(listOf<EnergyTypes?>(*EnergyTypes.entries.toTypedArray()))
    }
}
